src/people.rs

```diff
@@ -51,7 +58,93 @@ impl fmt::Debug for PersonId {
 // They may be defined with the define_person_property! macro.
 pub trait PersonProperty: Copy {
     type Value: Copy;
-    fn initialize(context: &Context, person_id: PersonId) -> Self::Value;
+    #[must_use]
+    fn is_derived() -> bool {
+        false
+    }
+    #[must_use]
+    fn dependencies() -> Vec<Box<dyn PersonPropertyHolder>> {
+        panic!("Dependencies not implemented");
+    }
+    fn compute(context: &Context, person_id: PersonId) -> Self::Value;
+    fn get_instance() -> Self;
+}
+
+type ContextCallback = dyn FnOnce(&mut Context);
+
+// The purpose of this trait is to allow us to store a vector of different PersonProperties
+// in an object safe way, and use them to emit change events.
+pub trait PersonPropertyHolder {
+    // Adds a callback to callback_vec which can later be called with a new context
+    // to emit a change event
+    fn add_event_callback(
+        &self,
+        context: &mut Context,
+        person: PersonId,
+        callback_vec: &mut Vec<Box<ContextCallback>>,
+    );
+    fn is_derived(&self) -> bool;
+    fn dependencies(&self) -> Vec<Box<dyn PersonPropertyHolder>>;
+    fn non_derived_dependencies(&self) -> Vec<TypeId>;
+    fn collect_non_derived_dependencies(&self, result: &mut HashSet<TypeId>);
+    fn property_type_id(&self) -> TypeId;
+}
+
+impl<T> PersonPropertyHolder for T
```

Comment:
As an aside I spent a while trying to figure out how to make this a supertrait to avoid the forwarders, but regrettably Copy appears to require Sized and so you can't make it object safe.